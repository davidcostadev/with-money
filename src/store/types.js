export const ADD_USER = 'ADD_USER';
export const TO_CURRENT_MONTH = 'TO_CURRENT_MONTH';
export const TO_PREVIUS_MONTH = 'TO_PREVIUS_MONTH';
export const TO_NEXT_MONTH = 'TO_NEXT_MONTH';
