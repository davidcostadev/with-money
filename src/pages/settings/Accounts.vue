<template>
  <div class="full-height">
    <md-toolbar>
      <router-link to="/settings">
        <md-button class="md-icon-button">
          <md-icon>keyboard_backspace</md-icon>
        </md-button>
      </router-link>
      <h3 class="md-title">Contas</h3>
    </md-toolbar>
    <md-content>
      <md-list v-for="account in accounts" :key="account.id">
        <md-list-item :to="`/settings/accounts/edit/${account.id}`">
          <span>{{account.name}}</span>
          <account-icon :value="account.type" />
        </md-list-item>
      </md-list>
    </md-content>

    <md-button
      class="md-fab md-primary menu-speed"
      to="/settings/accounts/new"
    >
      <md-icon>note_add</md-icon>
    </md-button>
  </div>
</template>

<script>
import Accounts from '../../services/accounts';
import AccountIcon from '../../components/AccountIcon';

export default {
  data() {
    return {
      accounts: [],
    };
  },
  components: { AccountIcon },
  methods: {
    async getAccounts() {
      const { data } = await Accounts.list();

      this.accounts = data;
    },
  },
  created() {
    this.getAccounts();
  },
};
</script>
