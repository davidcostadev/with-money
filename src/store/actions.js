import * as types from './types';

export default {
  addUser({ commit }, payload) {
    commit(types.ADD_USER, payload);
  },
  toNextMonth({ commit }) {
    commit(types.TO_NEXT_MONTH);
  },
  toCurrentMonth({ commit }) {
    commit(types.TO_CURRENT_MONTH);
  },
  toPreviousMonth({ commit }) {
    commit(types.TO_PREVIUS_MONTH);
  },
};
