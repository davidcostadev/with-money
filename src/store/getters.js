export default {
  user(state) {
    return state.user;
  },
  current_month(state) {
    return state.current_month;
  },
  state_month(state) {
    return state.state_month;
  },
  state_month_str(state) {
    return state.state_month_str;
  },
  next_month_str(state) {
    return state.next_month_str;
  },
  previous_month_str(state) {
    return state.previous_month_str;
  },
};
