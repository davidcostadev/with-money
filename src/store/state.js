import Moment from 'moment';

export default {
  user: null,
  current_month: Moment(),
  state_month: Moment(),
  state_month_str: Moment().format('MMM YY'),
  next_month_str: Moment().add(1, 'month').format('MMM YY'),
  previous_month_str: Moment().subtract(1, 'month').format('MMM YY'),
  flash_message: '',
};
