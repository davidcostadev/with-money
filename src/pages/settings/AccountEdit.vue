<template>
  <div class="full-height">
    <md-toolbar class="toolbar">
      <router-link to="/settings/accounts">
        <md-button class="md-icon-button">
          <md-icon>keyboard_backspace</md-icon>
        </md-button>
      </router-link>
      <h3 class="md-title" style="flex: 1">Conta: {{name}}</h3>
      <md-button id="delete" class="md-icon-button" @click="showConfirmDelete = true">
        <md-icon>delete</md-icon>
      </md-button>
    </md-toolbar>
    <div class="content-scroll">
      <account-form
        v-if="account"
        :name="name"
        :initalValue="initalValue"
        :type="type"
        :onSave="onSave"
      />
      <md-dialog-confirm
        :md-active.sync="showConfirmDelete"
        md-title="Deseja mesmo excluir essa transação?"
        md-confirm-text="Sim"
        md-cancel-text="Não"
        @md-cancel="showConfirmDelete = false"
        @md-confirm="onConfirm" />
    </div>
  </div>
</template>

<script>
import AccountForm from '../../components/forms/Account';
import Accounts from '../../services/accounts';

export default {
  components: {
    AccountForm,
  },
  data() {
    return {
      account: null,
      name: '',
      initalValue: '',
      type: 'out',
      showConfirmDelete: false,
    };
  },
  methods: {
    async onSave(data) {
      const account = await Accounts.update(this.$route.params.id, data);

      if (account.id) {
        this.$store.dispatch('showFlashMessage', 'Conta salva com sucesso!');
      }

      this.$router.push({
        path: '/settings/accounts',
      });
    },
    async getAccount() {
      const data = await Accounts.get(this.$route.params.id);

      this.name = data.name;
      this.initalValue = data.initalValue;
      this.type = data.type;

      this.account = data;
    },
    async onConfirm() {
      await Accounts.destroy(this.$route.params.id);

      this.$store.dispatch('showFlashMessage', 'Conta excluida com sucesso');

      this.$router.push({
        path: '/settings/accounts',
      });
    },
  },
  created() {
    this.getAccount();
  },
};
</script>
