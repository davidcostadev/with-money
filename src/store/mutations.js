import Moment from 'moment';
import * as types from './types';
import { updateStatesMont } from '../utils/state';

export default {
  [types.ADD_USER](state, payload) {
    state.user = payload;
  },
  [types.TO_CURRENT_MONTH](state) {
    state.state_month = state.current_month;
    updateStatesMont(state);
  },
  [types.TO_PREVIUS_MONTH](state) {
    state.state_month = Moment(state.state_month).subtract(1, 'month');
    updateStatesMont(state);
  },
  [types.TO_NEXT_MONTH](state) {
    state.state_month = Moment(state.state_month).add(1, 'month');
    updateStatesMont(state);
  },
};
