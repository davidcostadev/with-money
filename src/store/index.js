import Vue from 'vue';
import Vuex from 'vuex';
import Moment from 'moment';
import mutations from './mutations';
import actions from './actions';

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
    user: null,
    current_month: Moment(),
    state_month: Moment(),
    state_month_str: Moment().format('MMMM YY'),
    next_month_str: Moment().add(1, 'month').format('MMMM YY'),
    previous_month_str: Moment().subtract(1, 'month').format('MMMM YY'),
  },
  getters: {
    user(state) {
      return state.user;
    },
    current_month(state) {
      return state.current_month;
    },
    state_month(state) {
      return state.state_month;
    },
    state_month_str(state) {
      return state.state_month_str;
    },
    next_month_str(state) {
      return state.next_month_str;
    },
    previous_month_str(state) {
      return state.previous_month_str;
    },
  },
  mutations,
  actions,
});
