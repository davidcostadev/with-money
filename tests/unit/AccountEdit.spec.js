import { createLocalVue, shallowMount } from '@vue/test-utils';
import VueMaterial from 'vue-material';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
import AccountEdit from '../../src/pages/settings/AccountEdit';
import store from '../../src/store';
import Accounts from '../../src/services/accounts';

const localVueRender = createLocalVue();
const localVue = createLocalVue();

localVueRender.use(Vuex);
localVueRender.use(VueMaterial);
localVueRender.use(VueRouter);
localVue.use(Vuex);
localVue.use(VueMaterial);

const router = new VueRouter({
  name: 'AccountEdit',
  params: {
    id: 1,
  },
});

/* eslint global-require: "off" */
jest.mock('../../src/services/accounts', () => {
  const mockedAccounts = require('./__mocks__/accounts');

  return {
    update: jest.fn().mockResolvedValue(mockedAccounts.account1),
    get: jest.fn().mockResolvedValue(mockedAccounts.account1),
    destroy: jest.fn().mockResolvedValue(),
  };
});

describe('AccountEdit Component', () => {
  beforeAll(() => {
    global.localStorage = {
      getItem: jest.fn().mockReturnValue(JSON.stringify({
        id: '1',
        name: 'name',
      })),
    };
  });

  describe('Render', () => {
    it('default', () => {
      const wrapper = shallowMount(AccountEdit, { store, localVue: localVueRender, router });

      expect(wrapper.isVueInstance).toBeTruthy();
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Beravior', () => {
    it('onSave', async () => {
      const mocks = {
        $route: {
          name: 'AccountEdit',
          params: {
            id: 123,
          },
        },
        $router: {
          push: jest.fn(),
        },
      };
      const wrapper = shallowMount(AccountEdit, {
        localVue,
        store,
        router,
        mocks,
        stubs: ['router-link', 'router-view'],
      });

      await wrapper.vm.onSave({
        id: 1,
        name: 'name',
      });

      expect(Accounts.update).toBeCalled();
      expect(Accounts.update.mock.calls[0][0]).toBe(123);
      expect(Accounts.update.mock.calls[0][1]).toEqual({
        id: 1,
        name: 'name',
      });
      expect(mocks.$router.push.mock.calls[0][0]).toEqual({
        path: '/settings/accounts',
      });
    });

    it('onConfirm', async () => {
      const mocks = {
        $route: {
          name: 'AccountEdit',
          params: {
            id: 123,
          },
        },
        $router: {
          push: jest.fn(),
        },
      };
      const wrapper = shallowMount(AccountEdit, {
        propsData: {
          type: 'in',
        },
        localVue,
        store,
        router,
        mocks,
        stubs: ['router-link', 'router-view'],
      });

      await wrapper.vm.onConfirm({
        id: 1,
        name: 'name',
        type: 'out',
      });
      expect(Accounts.destroy).toBeCalled();
      expect(Accounts.destroy.mock.calls[0][0]).toBe(123);
      expect(mocks.$router.push).toBeCalled();
      expect(mocks.$router.push.mock.calls[0][0]).toEqual({
        path: '/settings/accounts',
      });
    });
  });
});
