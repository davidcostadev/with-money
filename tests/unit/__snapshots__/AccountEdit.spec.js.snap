// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountEdit Component Render default 1`] = `
<div>
  <mdtoolbar-stub
    mdelevation="4"
  >
    <router-link-stub
      event="click"
      tag="a"
      to="/settings/accounts"
    >
      <mdbutton-stub
        class="md-icon-button"
        type="button"
      >
        <mdicon-stub>
          keyboard_backspace
        </mdicon-stub>
      </mdbutton-stub>
    </router-link-stub>
     
    <h3
      class="md-title"
    >
      Conta: 
    </h3>
     
    <mdbutton-stub
      class="md-icon-button"
      id="delete"
      type="button"
    >
      <mdicon-stub>
        delete
      </mdicon-stub>
    </mdbutton-stub>
  </mdtoolbar-stub>
   
  <!---->
   
  <mddialogconfirm-stub
    mdcanceltext="Não"
    mdconfirmtext="Sim"
    mdtitle="Deseja mesmo excluir essa transação?"
  />
</div>
`;
